---
title: "Merqury Rising: Diploid QC Test"
author: "Richard J. Edwards"
date: "06/12/2023"
version: "v0.4.3"
output:
  html_document:
    #css: http://www.slimsuite.unsw.edu.au/stylesheets/slimhtml.css
    toc: true
    toc_float: true
    toc_collapsed: false
    toc_depth: 3
    number_sections: true
---

**NOTE:** This Rmd is currently for testing the diploid code only. The assembly comparisons are therefore not that meaningful. A dedicated diploid case study will be added in future releases.

# Introduction

[MerquryRising](https://github.com/slimsuite/MerquryRising) is a standalone RMarkdown file for parsing outputs from [Merqury](https://github.com/marbl/merqury) and generating some additional plots to help assess the requirements and/or consequences of duplicate purging in genome assemblies. Please see the accompanying [Tutorial.html](./Tutorial.html) for additional information about the rationale behind the program and its features. The GitHub README will also contain some information about running the script. If you are not getting the outputs you expect, please check or update the `merquryrising.config` configuration file, consisting of tab-delimited pairs of `setting` and `value`. 

The Merqury files needed are:

* A set of `*.spectra-cn.hist` files of read and assembly kmer counts.
* A corresponding set of `*.only.hist` files of assembly-only kmer counts.
* An optional read kmer `*.hist.ploidy` file given with `ploidyfile=FILE`. If this is a `*.hist` file, this will be used to set all the boundaries.

Input files are read from the directory specified by `merqurydir=PATH` (default:`merqury`) and are set by `histfiles=X` (wildcards allowed, default:`*.spectra-cn.hist`). A matching set of `*.only.hist` files are also expected. Human-friendly labels can be provided with a `merquryrising.fofn` file (`labels=FILE`) that consists of `label filename` as plain text. `filename` in this instance is the full name of the `*.spectra-cn.hist` file, excluding the path.

1. Default mode will compare each kmer histogram with the preceding one.
2. Reference mode will compare every kmer histogram with the same reference. (Either named, else the first file.)

To check the settings loaded from the config file, edit the Rmarkdown to have `debug=TRUE`. See the appendix and tutorial for more information about the MerquryRising workflow and classification definitions.

```{r rmd, echo=FALSE, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = FALSE, include = TRUE, warning=FALSE, message=FALSE)
# HISTORY
# - See main MerquryRising.Rmd

# TO DO
# [ ] : Update to match main Rmd but with diploid data.
# [ ] : Sort out the input data and update configs.

# NOTES
#!# Consider setting results='asis' to avoid wrapping in ```.
#?# I wonder if this is the solution for HTML output within loops?

#i# Set defaults for the Rmd implementation that might be different from the standalone version
# Set default settings. Can be over-ridden with merquryrising.config.
override <- c('fullrun=FALSE','rdir=.','basefile=mqrdip','config=tutorial.config','labels=tutorial.fofn',
              'debug=FALSE','dev=TRUE','makexlsx=FALSE')
source('./merquryrising.R')
```

# Input Data

The main input for MerquryRising is the set of `*.spectra-cn.hist` and corresponding `.only.hist` files, with optional `qv/*.merqury.qv` and `stats/*.merqury.completeness.stats` files. The latter are matched on the `name` field, which is parsed from the `*.spectra-cn.hist` and split on `.merqury.`. (See the example data if unclear.)

Optionally, assembly names will be mapped from the `labels=FILE`, otherwise the label will be the "G" number, `G1` to `Gn`. (See Tutorial for field descriptions.)

## Input files and assemblies

The following input files and assemblies have been recognised for this run:

```{r set_files}
#i# Creation is now handled in the R script.
# D$adb <- setInputFiles()
#i# Summarise the input data table
if(settings$debug){
  kable(D$adb, "html", caption = "Input genome assemblies and QV/completeness stats.", align = "l") %>% kable_styling("striped", full_width = F)  
}else{
  kable(D$adb %>% select(-file,-ofile), "html", caption = "Input genome assemblies and QV/completeness stats.", align = "l") %>% kable_styling("striped", full_width = F)
}
```


## Setting the ploidy boundaries

The next step is set the ploidy boundaries that establish `rfreq` categories as `low`, `haploid`, `diploid` or `high` frequencies. Assembly-only kmers are classified as `only`. 

The original iterations of this script used the `boundary` values loaded from the `ploidy` file, but these appear to be quite conservative and result in a lot of kmers being rated as over- or under-purged based on the merqury plots themselves, with a lot of the main distributions broader than the values loaded. Ploidy boundaries are therefore set by default to be determined from the histograms themselves. For consistent boundaries, provide a `*.hist` file to the `ploidy=FILE` setting.

When `*.hist` files are used (assembly-specific or general), the diploid peak is first identified as the kmer density peak for rfreq values up to 1000. The boundaries are then calculated based on this `diploid` peak:

* `haploid` frequency is set to half `diploid`.
* The `low` boundary is set to the _maximum_ value of 5, or 0.25 `haploid`. Upto this value of `rfreq` (raw read kmer frequency), kmers are rated `low`. This is plotted as a solid black line on the Comparative Merqury plots.
* The `mid` boundary is set to halfway between `haploid` and `diploid`. Raw kmer frequencies above the `low` boundary and upto the `mid` boundary `hap`. This is plotted as a solid red line on the Comparative Merqury plots, with the haploid peak a dotted red line.
* The `high` boundary is set to 1.5 `diploid`. Raw kmer frequencies above the `hap` boundary and upto the `high` boundary are rated `dip`. This is plotted as a solid blue line on the Comparative Merqury plots, with the diploid peak a dotted blue line.
* Above the `high` boundary, kmers are rated as `high`.

**NOTE:** For very heterozygous genomes, where the haploid peak exceeds the diploid peak, the diploid peak calculation will break. For these, it is advised to make and load a `*.ploidy` file to use for all assemblies.

**NOTE:** The classification percentages are ultimately defined by the ploidy boundaries, which do not fully capture the extent of each sub-distribution of kmers. As a consequence, there will always be a degree of mis-assigned kmers. For this reason, percentages are best considered as comparative values rather than absolute values.


```{r load_ploidy}
#i# Optional loading of the ploidy file or generic ploidy boundaries are now handled in the main R script.

if(nrow(D$ploidy) > 0){
  kable(D$ploidy, "html", caption = "Ploidy boundaries for all assemblies, loaded from ploidy file.", align = "l") %>% kable_styling("striped", full_width = F)
}
if(nrow(D$boundary) > 0){
  kable(D$boundary, "html", caption = "Ploidy boundaries for all assemblies.", align = "l") %>% kable_styling("striped", full_width = F)
}
```

## Load and integrate kmer frequency data








## Load hist tables

Generate `Table1` from the hist files.

```{r loadhist}
#i# All the tables are now elements of the D list, e.g. D$Tab1, D$Tab2 etc.
# - Load the `*.spectra-cn.hist` data tables into tibbles with 3 fields: `afreq` (kmer frequency in assembly), `rfreq` (kmer frequency in reads) and `knum` (number of different kmers).
D <- makeTab1(D)  #i# Main code moved to the Rscript.

# Display boundaries loaded from the individual distributions
if(nrow(D$boundary) > 0){
  kable(D$boundary, "html", caption = "Ploidy boundaries.", align = "l") %>% kable_styling("striped", full_width = F)
}
#logWrite(names(D))
```



# Processing


## Reference-normalised scores

Generate `Table2` from `Table1`, with all values relative to reference.


```{r Table2}
# - Convert `Table1` into a version of all values relative to reference. [`Table2`]
#i# If settings$reference is `default`, will use the previous genome as the reference in each case
#i# Otherwise, will compare all to a specific reference
D <- makeTab2(D)
```

At this point, we have two "wide" tables of the kmer counts for each genome that belong to different assembly (`afreq`) and raw (`rfreq`) frequency classes. `Table1` has absolute values, whilst `Table2` has values relative to the reference. 


## Classification table

Generate `Table3`. This is a "long" version of Table 1, with the addition of a `knum` field. `afreq` and `rfreq` are then converted to the categories based on `afreq` and ploidy. Assembly-only kmers are arbitrarily given haploid or diploid kmer frequencies for the purposes of calculating the total kmer percentage for these categories.

```{r Table3}
# - Convert `Table1` into a long version [`Table3`] -> `knum` and `assembly`.
D <- makeTab3(D)
```

`Table3` should now have haploid- and diploid-specific ratings.

```{r palettestuff}
#i# Moved to Rscript.
```

# Results

## Purge classification



```{r Table4, fig.width=6, fig.height=3}
# - Create a `Table4` from `Table3` with the `purge` rating categories
#i# Rounding to digits seems more useful than sigfig
D <- makeTab4(D)
kable(D$Tab4w,"html", caption = "Purge rating (percentage).", align = "l") %>% kable_styling("striped", full_width = F)
saveToTSV(D$Tab4w %>% mutate_at(vars(only:qv), as.numeric),"rating","Purge rating (percentage)")

# Generate and save the main purge plot
p <- purgePlot(D$Tab4)
print(p)
```


```{r Table5, fig.width=6, fig.height=3}
# - Create a `Table5` from `Table3` with the detailed `purge` classes categories
#i# Rounding to digits seems more useful than sigfig
D <- makeTab5(D)
kable(D$Tab5w, "html", caption = "Kmer classification (percentage).", align = "l") %>% kable_styling("striped", full_width = F)
saveToTSV(D$Tab5w %>% mutate_at(vars(only:missing), as.numeric),"class","Kmer classification (percentage)")

# Generate and save the purge class plot
p <- classPlot(D$Tab5)
print(p)
```

## Relative purge classification

```{r relpurge}
#i# Use Tab4w and Tab5w for difference versus reference tables and plots (next section)
# - Reshape `Table5` and `Table4` wide and convert into difference versus reference. Generate difference plots.
D <- makeRelPurge(D)


kable(D$Tab4relw,"html", caption = "Relative purge rating (percentage).", align = "l") %>% kable_styling("striped", full_width = F)
p <- purgePlot(D$Tab4rel,ptype="relrating")
print(p)
savePlot(p,nrow(D$Tab4relw),ptype="relrating")

kable(D$Tab5relw, "html", caption = "Relative Kmer classification (percentage).", align = "l") %>% kable_styling("striped", full_width = F)
p <- classPlot(D$Tab5rel,ptype="relclass")
print(p)
savePlot(p,nrow(D$Tab5relw),ptype="relclass")
```





```{r merquryPlot}
#i# This has been moved to the Rscript
```

## Comparative Merqury plots

These plots have been converted into relative values compared to the reference assembly. The reference assembly will have the full Merqury plot. Note that the read-only kmers (`afreq`=0) may be truncated.

```{r plotsVsRef, eval=FALSE, include=FALSE, results='asis'}
### ~ Relative change data/plots ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ###
#i# This function generates the relative change output for assembly i of D$adb
#i# This initial version does not use the two-panel plotting.
relChangeV1 <- function(i){
  pref = FALSE
  pTab <- D$Tab2
  ytitle <- "kmer count shift" 
  if(settings$reference == "default"){
    if(i < 2){
      pref <- TRUE
    }
    refname <- D$adb$name[i-1]
  }
  if(settings$reference == D$adb$name[i]){
    pref <- TRUE
  }
  if(settings$reference %in% D$adb$name){
    refname <- settings$reference
  }
  if(pref){
    pTab <- D$Tab1 #%>% filter(rfreq > 1)
    ytitle <- "No. distinct kmers"
  }
  G <- D$adb$G[i]
  # Pull out the data to plot for genome G: rename the G field and dump the rest
  dat <- pTab %>% rename(knum=!!G) %>% select(afreq,rfreq,knum)
  p <- merquryPlot(dat,ytitle,G)
  cat(paste('###',D$adb$name[i],'\n\n'))
  
  #print(kable(D$adb[i,], "html", caption = D$adb$name[i], align = "l") %>% kable_styling("striped", full_width = T))
  
  cat(paste('\n\n'))
  if(! pref){
    #!# Add the relative changes from Tab4w/Tab4relw and Tab5w/Tab5relw
    ai <- D$adb$label[i]
    aj <- D$adb[D$adb$name == refname,]$label[1]

    reltab <- bind_rows(D$Tab4w %>% filter(assembly==!!aj),
                        D$Tab4relw %>% filter(assembly==!!ai),
                        D$Tab4w %>% filter(assembly==!!ai)) %>% select(-G)
    reltab$assembly[2] <- "change"
    kcap <- paste0('Relative purge rating of ',ai,' versus ',aj)
    print(kable(reltab, "html", caption = kcap, align = "l") %>% kable_styling("striped", full_width = T))

    reltab <- bind_rows(D$Tab5w %>% filter(assembly==!!aj),
                    D$Tab5relw %>% filter(assembly==!!ai),
                    D$Tab5w %>% filter(assembly==!!ai)) %>% select(-G)
    reltab$assembly[2] <- "change"
    kcap <- paste0('Relative kmer classes of ',ai,' versus ',aj)
    print(kable(reltab, "html", caption = kcap, align = "l") %>% kable_styling("striped", full_width = T))
        
    cat(paste0('\n\nRelative change of ',D$adb$name[i],' versus ',refname,':\n\n'))
  }else{
    ai <- D$adb$label[i]

    reltab <-D$Tab4w %>% filter(assembly==!!ai)
    kcap <- paste0('Purge rating of ',ai)
    print(kable(reltab, "html", caption = kcap, align = "l") %>% kable_styling("striped", full_width = T))

    reltab <- D$Tab5w %>% filter(assembly==!!ai)
    kcap <- paste0('kmer classes for ',ai)
    print(kable(reltab, "html", caption = kcap, align = "l") %>% kable_styling("striped", full_width = T))
    cat(paste0('\n\nMerqury plot for ',D$adb$name[i],'.\n\n'))
  }
  if(max(dat$knum) > 0){
    print(p)
  }else{
    cat(paste0('No change!\n\n'))
  }
  #cat(paste0('\n\n**Plot of relative change versus',refname,'.**\n\n'))
  cat('\n\n')
  return()
}
```

```{r plotsVsRef2, results='asis', fig.width=12, fig.height=4}
### ~ Relative change data/plots ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ###
#i# This function generates the relative change output for assembly i of D$adb
#i# This  version uses two-panel plotting.
relChange <- function(i){
  #># First, set the reference
  pref = FALSE
  if(settings$reference == "default"){
    if(i < 2){
      pref <- TRUE
    }
    refname <- D$adb$name[i-1]
  }
  if(settings$reference == D$adb$name[i]){
    pref <- TRUE
  }
  if(settings$reference %in% D$adb$name){
    refname <- settings$reference
  }
  #># Set the variables for this assembly version
  G <- D$adb$G[i]
  Gname <- D$adb$name[i]
  Glabel <- D$adb$label[i]
  Gbase <- D$adb$base[i]
  
  #># Output summary information/tables
  cat(paste('###',Glabel,'\n\n'))
  if(! pref){
    cat(paste0("Merqury kmer statistics and relative purge stats for ",Gname," (`",Gbase,".*`)."))  
    cat(paste('\n\n'))
    ai <- D$adb$label[i]
    aj <- D$adb[D$adb$name == refname,]$label[1]

    reltab <- bind_rows(D$Tab4w %>% filter(assembly==!!aj),
                        D$Tab4relw %>% filter(assembly==!!ai),
                        D$Tab4w %>% filter(assembly==!!ai)) %>% select(-G)
    reltab$assembly[2] <- "change"
    kcap <- paste0('Relative purge rating of ',ai,' versus ',aj)
    print(kable(reltab, "html", caption = kcap, align = "l") %>% kable_styling("striped", full_width = T))

    reltab <- bind_rows(D$Tab5w %>% filter(assembly==!!aj),
                    D$Tab5relw %>% filter(assembly==!!ai),
                    D$Tab5w %>% filter(assembly==!!ai)) %>% select(-G)
    reltab$assembly[2] <- "change"
    kcap <- paste0('Relative kmer classes of ',ai,' versus ',aj)
    print(kable(reltab, "html", caption = kcap, align = "l") %>% kable_styling("striped", full_width = T))
        
    cat(paste0('\n\nRelative change of ',D$adb$name[i],' versus ',refname,':\n\n'))
  }else{
    cat(paste0("Merqury kmer statistics for ",Gname," (`",Gbase,".*`). (No reference.)"))  
    cat(paste('\n\n'))
    ai <- D$adb$label[i]

    reltab <-D$Tab4w %>% filter(assembly==!!ai)
    kcap <- paste0('Purge rating of ',ai)
    print(kable(reltab, "html", caption = kcap, align = "l") %>% kable_styling("striped", full_width = T))

    reltab <- D$Tab5w %>% filter(assembly==!!ai)
    kcap <- paste0('kmer classes for ',ai)
    print(kable(reltab, "html", caption = kcap, align = "l") %>% kable_styling("striped", full_width = T))
    cat(paste0('\n\nMerqury plot for ',D$adb$name[i],'.\n\n'))
  }
  
  # Two panel plots
  # Pull out the data to plot for genome G: rename the G field and dump the rest
  pTab <- D$Tab1
  ytitle <- "No. distinct kmers"
  dat <- pTab %>% rename(knum=!!G) %>% select(afreq,rfreq,knum)
  p1 <- merquryPlot(dat,ytitle,G)
  # Second panel
  pTab <- D$Tab2
  ytitle <- "kmer count shift" 
  dat <- pTab %>% rename(knum=!!G) %>% select(afreq,rfreq,knum)
  p2 <- merquryPlot(dat,ytitle,G)
  # Two-panel plot with patchwork  
  p <- p1 + p2
  print(p)
  cat('\n\n')
  savePlot(p,ptype=Glabel)
  cat('\n\n')
  return()
}

#i# Generate the relative kmer output for each assembly
for(i in 1:nrow(D$adb)){
  relChange(i)
}
```

# Appendix

## Merqury Rising workflow

The general `MerquryRising` workflow is as follows:

- Run Merqury.
- Establish list of input assembly files and build alias list (`G1` to `Gn`).
- Load the read kmer frequency boundaries from the ploidy file: `ploidy`, `depth`, `boundary`.
- Load the `*.spectra-cn.hist` data tables into tibbles with 3 fields: `afreq` (kmer frequency in assembly), `rfreq` (kmer frequency in reads) and `knum` (number of different kmers).
- Convert the `read-only` kmers into `0` and make sure all fields are integers.
- Add the `*.only.hist` data to each table.
- Join all the count tables by `afreq` and `rfreq` to give a field per input file. (`G1` to `Gn`). [`Table1`]
- Convert `Table1` into a version of all values relative to reference. [`Table2`]
- Convert `Table1` into a long version [`Table3`] -> `knum` and `assembly`.
- Convert `Table3` `rfreq` into categories based on `ploidy`: `low`, `hap`, `dip`, `high`.
- Create a `Table3` `class` field, based on `afreq` and `rfreq` (see below)
- Collapse by `class` and sum up the `knum`
- Create a `Table4` from `Table3` with the `purge` categories (see below)
- Generate a horizontal stacked bar plot of each assembly: noise, only, lowQ, duplicate, alternate, haploid, diploid, repeats, collapsed, missing [`Table3`]
- Generate a horizontal stacked bar plot of each assembly: noise, only, under, good, over [`Table4`]
- Generate a kmer difference plot for each assembly versus the reference with afreq of 0, 1, 2, 3+ [`Table2`]
- Reshape `Table3` and `Table4` wide and convert into difference versus reference. Generate difference plots.

**NOTE:** The `only` class could indicate assembly errors, or it could indicate missing kmers due to sequencing biases. As such, they are excluded from the classification, as are the `noise` kmers, which are suspected read errors or contamination.

Please see the Tutorial for more details.

## Kmer classes

Kmer classes will be assigned based on the read frequency (`rfreq`) and kmer frequency (`kfreq`), ploidy boundaries (see Tutorial and below), and whether the assembly is haploid or diploid. Classification is first based on more specific categories, which are then grouped into broader purge categories. The `dipclass` and `dippurge` ratings are used for diploid assemblies.

```{r dipkmerclasses}
#i# Definition now part of the merquryrising.R script.
kable(D$kclassdb, "html", caption = "MerquryRising Kmer frequency classes.", align = "c") %>% kable_styling("striped", full_width = F)
```

### Kmer classification

MerquryRising uses the following kmer classes:

* **only** = kmers found only in the assembly and not the raw reads. These could represent errors that have been introduced (e.g. by data from different, less accurate, sequencing technology), but they could also represent missing kmers due to biases in the raw reads.
* **noise** = kmers found only in the raw data and at `low` frequency. These are most likely to be base-calling **Read Errors**, or contamination.
* **lowQ** = kmers that are found only at `low` abundance in the raw data but also in the assembly indicate **Low Quality** regions and possible contamination.
* **duplicate** = kmers found at `haploid` frequency in the raw data but appearing more than once in the assembly are flagged as probable **Duplicate** regions. kmers at `diploid` frequency found more than once in a haploid assembly, or more than twice in a diploid assembly, are also flagged as `Duplicate`. These could represent scenarios where a haploid assembly contains both alleles, or where both haplotypes contain a hemizygous region that is only present in one haplotype.
* **alternate** = `haploid` frequency kmers that are not found in a haploid assembly are assumed to be from the **Alternate Haplotype**. (Where the assembly has been phased, can be confirmed by running Merqury on the combined diploid assembly of both haplotypes.)
* **haploid** = kmers found at `haploid` frequency in the raw data but and once in the assembly are flagged as **Haplotype Specific** kmers (or **Haploid kmers**).
* **diploid** = kmers found at `diploid` frequency in the raw data but and once in a haploid assembly, or twice in a diploid assembly, are flagged as **Diploid kmers**.
* **repeats** = kmers found at `high` frequency in the raw data and three or more times in the assembly are flagged as **Repeats**. These are not obviously collapsed but it is not possible to assess them for under- or over-representation based on these data. (Note that rDNA repeats will normally be in this category but be collapsed.)
* **collapsed** = kmers found at `high` frequency in the raw data but only one or two times in the assembly are flagged as **Collapsed** repeats that are under-represented in the assembly.
* **missing** = kmers found at `diploid` or `high` frequency in the raw data but missing from the assembly are flagged as **Missing**. For diploid assemblies, kmers found at `haploid` frequency in the raw data but missing from the assembly are also flagged as **Missing**.

### Purge Ratings

Classes are grouped into purge categories based on the observed assembly frequency (`afreq`) versus the expectation derived from the raw data (`rfreq`) :

* **only** = Possible base errors in the assembly, or missing raw read kmers due to sequencing biases etc. Do not inform on duplicate purging status. (Class: `only`)
* **n/a** = Errors or contamination in the raw data. Do not inform on duplicate purging status. (Class: `noise`)
* **under** = Under-purged kmers that should not be in the assembly. (Class: `lowQ`, `duplicate`)
* **good** = kmers where the assembly frequency is consistent with the raw read frequency (Class: `alternate`, `haploid`, `diploid`)
* **neutral** = Repeat kmers that are not obviously collapsed. Merqury does not report assembly frequencies above 4, making it impossible to judge whether many repetitive kmers are over- or under-represented. (Class: `repeats`)
* **over** = Over-purged kmers that appear to be lacking occurrences in the assembly, given their raw read frequency. (Class: `collapsed`, `missing`)

**NOTE:** The `only` class could indicate assembly errors, or it could indicate missing kmers due to sequencing biases. As such, they are excluded from the classification, as are the `noise` kmers, which are suspected read errors or contamination.




```{r planning, eval=FALSE, include=FALSE}
#i# Future updates should incorporate BUSCO and DepthKopy (AGRatio) outputs for comparison.


```